src/Sema.zig

```diff
@@ -5512,16 +5512,156 @@ fn zirArrayCat(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) CompileEr
+    if (t.zigTypeTag() == .Array) {
+        info.len = t.arrayLen();
+        if (t.tag() == .array_u8) {
+            info.array_elem_type = Type.initTag(.u8);
+        } else if (t.tag() == .array_u8_sentinel_0) {
+            info.array_elem_type = Type.initTag(.u8);
+            info.sent = Value.initTag(.zero);
+        } else if (t.tag() == .array) {
+            const payload = t.castTag(.array).?.data;
+            info.array_elem_type = payload.elem_type;
+        } else if (t.tag() == .array_sentinel) {
+            const payload = t.castTag(.array_sentinel).?.data;
+            info.array_elem_type = payload.elem_type;
+            info.sent = payload.sentinel;
+        }
+    } else if (t.zigTypeTag() == .Pointer) {
+        const ptrinfo = t.ptrInfo().data;
+        if (ptrinfo.pointee_type.zigTypeTag() != .Array) return null;
+        if (ptrinfo.size != .One) return null;
+        return getArrayCatInfo(ptrinfo.pointee_type);
+    } else return null;
+    return info;
 }
 
 fn zirArrayMul(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) CompileError!Air.Inst.Ref {
     const tracy = trace(@src());
     defer tracy.end();
 
-    _ = inst;
-    return sema.mod.fail(&block.base, sema.src, "TODO implement zirArrayMul", .{});
+    const inst_data = sema.code.instructions.items(.data)[inst].pl_node;
+    const extra = sema.code.extraData(Zir.Inst.Bin, inst_data.payload_index).data;
+    const lhs = sema.resolveInst(extra.lhs);
+    const rhs = sema.resolveInst(extra.rhs);
+    const lhs_ty = sema.typeOf(lhs);
+    const lhs_src: LazySrcLoc = .{ .node_offset_bin_lhs = inst_data.src_node };
+    const rhs_src: LazySrcLoc = .{ .node_offset_bin_rhs = inst_data.src_node };
+
+    // In `**` rhs has to be comptime-known, but lhs can be runtime-known
+    const tomulby = try sema.resolveInt(block, rhs_src, extra.rhs, Type.initTag(.usize));
+    const mulinfo = getArrayCatInfo(lhs_ty) orelse
+        return sema.mod.fail(&block.base, lhs_src, "expected array, found '{}'", .{lhs_ty});
+
+    const final_len = std.math.mul(u64, mulinfo.len, tomulby) catch return sema.mod.fail(&block.base, rhs_src, "operation results in overflow", .{});
+    if (try sema.resolveDefinedValue(block, lhs_src, lhs)) |lhs_val| {
+        if (lhs_val.tag() == .empty_array) return lhs;
+        if (lhs_val.castTag(.decl_ref)) |lhs_sub| {
+            const lhs_sub_val = lhs_sub.data.val;
+            var anon_decl = try block.startAnonDecl();
+            defer anon_decl.deinit();
+
+            const final_ty = if (mulinfo.sent) |sent|
+                try Type.Tag.array_sentinel.create(anon_decl.arena(), .{ .len = final_len, .elem_type = mulinfo.array_elem_type, .sentinel = sent })
+            else
+                try Type.Tag.array.create(anon_decl.arena(), .{ .len = final_len, .elem_type = mulinfo.array_elem_type });
+
+            if (lhs_sub_val.tag() == .empty_array) return rhs;
```

Comment:
Should that be moved right after the declaration of ´lhs_sub_val´, you are neither using `anon_decl` nor `final_ty` here

Reply:
I don't see why, I put it there so eventually, it could be moved out of the block as more things could use it when arraycat works at runtime also.